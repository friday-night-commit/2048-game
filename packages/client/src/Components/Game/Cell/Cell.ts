import { generateValue } from '../utils/getValue'
import { ThemeProvider } from '@material-tailwind/react'
import value = ThemeProvider.propTypes.value

export type Position = {
  x: number,
  y: number
}


export class Cell {
  protected readonly _value: number
  protected _position: Position
  protected context: CanvasRenderingContext2D;

  constructor(context: CanvasRenderingContext2D, position: Position) {
    // TODO Игра управляет ячейками, сама ячейка не знает ничего, кроме значения и позиции
    this.context = context;
    this._value = generateValue();
    this._position = position;
  }

  get value(): number {
    return this._value
  }

  setPosition(newPosition: Position) {
    this._position = newPosition;
  }

  get position(): Position {
    return this._position
  }

  changeColor() : string {
    switch (this.value) {
      case 2:
        return 'rgb(224, 222, 220)';
      case 4:
        return 'rgb(209, 206, 203)';
      case 8:
        return 'rgb(194, 190, 186)';
      case 16:
        return 'rgb(179, 174, 169)';
      case 32:
        return 'rgb(164, 158, 152)';
      case 64:
        return 'rgb(149, 142, 135)';
      case 128:
        return 'rgb(255, 216, 107)';
      case 256:
        return 'rgb(255, 186, 0)';
      case 512:
        return 'rgb(211, 143, 74)';
      case 1024:
        return 'rgb(205, 127, 50)';
      case 2048:
        return 'rgb(237, 111, 49)';
      case 4096:
        return 'rgb(255, 31, 31)';
      default:
        return 'gray';
    }
  }

  render() {
    this.context.fillStyle = this.changeColor();
    const { x, y } = this.position
    this.context.fillRect(x*75,y*75,75,75);
    this.context.textBaseline = 'middle';
    this.context.fillStyle = 'black';
    this.context.textAlign = 'center';
    this.context.fillText(String(this.value), x*75+37, y*75+35);
  }
}