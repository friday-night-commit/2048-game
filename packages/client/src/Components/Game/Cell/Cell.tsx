import { Engine } from '../Engine/Engine'

export type Position = {
  x: number,
  y: number
}


export class Cell {
  protected readonly value: number
  protected position: Position
  protected context: CanvasRenderingContext2D;

  constructor(context: CanvasRenderingContext2D, position: Position) {
    // TODO Игра управляет ячейками, сама ячейка не знает ничего, кроме значения и позиции
    this.context = context;
    this.value = this.generateValue();
    this.position = position;
  }

  getValue(): number {
    return this.value
  }

  setPosition(newPosition: Position): void {
    this.position = newPosition;
  }

  getPosition(): Position {
    return this.position
  }

  private generateValue(): number {
    return Math.random() > 0.75 ? 4 : 2
  }

  render() {
    this.context.fillStyle = 'gray';
    const { x, y } = this.getPosition()
    this.context.fillRect(x*75,y*75,75,75);
    this.context.font = "20px Arial";
    this.context.textBaseline = 'middle';
    this.context.fillStyle = 'black';
    this.context.textAlign = 'center';
    this.context.fillText(String(this.value), x*75+37, y*75+35);
  }

  // getCellPosition() {
  //   const emptyCoordinates = this.engine.getEmptyMatrixCoordinates()
  //   const randomCoordinateIndex = this.engine.generateRandom(0, emptyCoordinates.length - 1)
  //
  //   return emptyCoordinates[randomCoordinateIndex]
  // }

  // setPosition(newPosition: Position) {
  //   const oldPosition = this.position
  //   this.position = newPosition
  //   this.cellEl = this.engine.reDrawCellElem(oldPosition, this.position, this.context, this.value)
  //   this.engine.addCellToMatrix(this)
  //   this.engine.removeCellFromMatrix(oldPosition)
  // }
  //

  //
  // move(moveDirection: Direction)
  // {
  //   const matrix = this.getMatrix();
  //   const { x, y } = this.getPosition();
  //
  //   switch (moveDirection) {
  //     case "left": {
  //       this.engine.findCellHorizontalPosition({
  //         cell: this,
  //         x,
  //         y,
  //         matrix,
  //         startCondition: (x: number) => !x,
  //         endCondition: (currentX: number) => currentX > 0,
  //         changeMethod: (currentX: number) => currentX - 1
  //       });
  //       break;
  //     }
  //     case "right": {
  //       this.engine.findCellHorizontalPosition({
  //         cell: this,
  //         x,
  //         y,
  //         matrix,
  //         startCondition: (x: number) => x === 3,
  //         endCondition: (currentX: number) => currentX < 3,
  //         changeMethod: (currentX: number) => currentX + 1
  //       });
  //       break;
  //     }
  //    }
  //}
}