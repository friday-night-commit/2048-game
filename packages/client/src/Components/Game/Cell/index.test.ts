import Cell from '.';

describe('[SNAPSHOTS] Cell class', () => {
  let canvas: HTMLCanvasElement, ctx: CanvasRenderingContext2D, cell: Cell;

  beforeEach(() => {
    canvas = document.createElement('canvas');
    ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
  });

  it('should have proper public attributes', () => {
    cell = new Cell({ x: 0, y: 0 }, 20, 2);
    expect(cell.value).toBe(2);
    expect(cell.position).toStrictEqual({ x: 0, y: 0 });
  });

  it('should draw a cell with specified parameters', () => {
    cell = new Cell({ x: 0, y: 0 }, 20, 2);
    cell.render(ctx);

    // @ts-ignore - __getEvents comes from jest-canvas-mock package
    const events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
  });

  it('should draw a cell with another specified value', () => {
    cell = new Cell({ x: 0, y: 0 }, 20, 4);
    cell.render(ctx);

    // @ts-ignore - __getEvents comes from jest-canvas-mock package
    const events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
  });
});

describe('[MOCKS] Cell class', () => {
  const mockFillRect = jest.fn();
  const mockFillText = jest.fn();

  beforeAll(() => {
    const createElement = document.createElement.bind(document);

    jest
      .spyOn(document, 'createElement')
      .mockImplementation(
        (tagName: string, _options?: ElementCreationOptions) => {
          if (tagName === 'canvas') {
            return {
              ...createElement(tagName),
              getContext: jest.fn(() => {
                return {
                  fillRect: mockFillRect,
                  fillText: mockFillText,
                };
              }),
            };
          }
          return createElement(tagName);
        }
      );
  });

  beforeEach(() => {
    mockFillRect.mockReset();
    mockFillText.mockReset();
  });

  it('should draw cell with specified parameters', async () => {
    const canvas: HTMLCanvasElement = document.createElement('canvas');
    const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
    const cell = new Cell({ x: 0, y: 0 }, 20, 4);
    cell.render(ctx);

    expect(mockFillRect.mock.calls[0]).toEqual([0, 0, 20, 20]);
    expect(mockFillText.mock.calls[0]).toEqual(['4', 10, 10]);
    expect(ctx.font).toBe('20px Arial');
    expect(ctx.textBaseline).toBe('middle');
    expect(ctx.fillStyle).toBe('black');
    expect(ctx.textAlign).toBe('center');
  });

  afterAll(() => {
    jest.clearAllMocks();
  });
});
