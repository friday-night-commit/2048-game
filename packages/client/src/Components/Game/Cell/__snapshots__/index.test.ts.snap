// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOTS] Cell class should draw a cell with another specified value 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#ede0c8",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "20px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "4",
      "x": 10,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`[SNAPSHOTS] Cell class should draw a cell with specified parameters 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#eee4da",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "20px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "2",
      "x": 10,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;
